Move sshd_config to sshd_config.d

rrotter opened the pull request:
### `sshd_config` / `ssh_config`
- add default distro versions of sshd_config, ssh_config to revert previous changes
- move our customizations to drop-ins in sshd_config.d, ssh_config.d
  - remove non-customizations (don't need to set settings that are already set the distro default config, unless they are very important)

### `pam.d/sshd`
- drop-in not supported here, relocate w/ `dpkg-divert`

Also see: #882, which removes the temporary code added here.

rrotter committed:
add dist versions of sshd_config file

rrotter committed:
mv sshd_config to sshd_config.d

rrotter committed:
add dist versions of ssh_config file

rrotter committed:
mv ssh_config to ssh_config.d

- replace ssh_config with distro default
- install our customizations in a drop-in .conf file

rrotter committed:
add dist versions of /etc/pam.d/sshd

rrotter committed:
divert pam.d/sshd

Since we can't replace this with a drop-in, tell dpkg to manage
/etc/pam.d/sshd-defaults *instead* as though it were /etc/pam.d/sshd.

daaang commented on files/default/bookworm/etc/ssh/ssh_config:
```diff
@@ -0,0 +1,53 @@
+#   RekeyLimit 1G 1h
+#   UserKnownHostsFile ~/.ssh/known_hosts.d/%k
+    SendEnv LANG LC_*
+    HashKnownHosts yes
```
HashKnownHosts isn't my favorite, and it seems especially silly for our servers, but I'm guessing this is just copying in debian defaults (and I don't dislike anywhere near enough to say "let's don't do defaults").

rrotter replied:
Good catch. I strongly agree with quashing this. Almost anything here can be overridden in a drop in. I'll add that.

rrotter replied:
Fixed in last push.

daaang commented on files/default/trixie/etc/ssh/sshd_config:
```diff
@@ -0,0 +1,124 @@
+#	X11Forwarding no
+#	AllowTcpForwarding no
+#	PermitTTY no
+#	ForceCommand cvs server
```
Yeah I'm not running diffs on these. I'm just assuming you copied in three default config files bundled with each of the four distro/versions we use.

rrotter replied:
Yep. Started each in containers, installed sshd, copied the respective files out.

daaang commented on manifests/profile/networking/sshd.pp:
```diff
@@ -26,26 +26,49 @@
+    notify => Service['sshd'],
+  }
+
+  file { '/etc/ssh/sshd_config.d/50-lit.conf':
```
Makes sense.

daaang commented on manifests/profile/networking/sshd.pp:
```diff
@@ -26,26 +26,49 @@
     notify  => Service['sshd'],
   }
 
-  concat { '/etc/ssh/ssh_config': }
```
Ah I was wondering where this was coming from.

daaang commented on manifests/profile/bolt.pp:
```diff
@@ -60,4 +60,6 @@
     }
   }
+
+  concat { '/etc/ssh/ssh_config.d/50-bolt.conf': }
```
Any reason for the concat with a single fragment instead of just a file? I'm assuming it was a concat previously in order to span profiles.

rrotter replied:
Look up to line 50, looks like a loop to me.

daaang commented on manifests/profile/networking/sshd.pp:
```diff
@@ -26,26 +26,49 @@
+    timeout => 30,
+    command => '/usr/bin/dpkg-divert --no-rename --divert /etc/pam.d/sshd-defaults --add /etc/pam.d/sshd',
+    before  => File['/etc/pam.d/sshd-defaults'],
+    require => Exec['divert pam.d/sshd'],
```
Ok.

1. If `/etc/pam.d/sshd-defaults` does not exist,
2. We run `dpkg-divert --rename --divert /etc/pam.d/sshd-defaults --add /etc/pam.d/sshd` and then
3. If `dpkg-divert --truename /etc/pam.d/sshd` is not `/etc/pam.d/sshd-defaults`
4. We run `dpkg-divert --no-rename --divert /etc/pam.d/sshd-defaults --add /etc/pam.d/sshd`

I read the man page, but I don't get what a divert is, or why we need to do it, or why the above sequence is important.

daaang replied:
Wait no, this is incomplete, and I'm guessing the above step 3 will prevent step 4 from ever happening, which is the point.

1. If `/etc/pam.d/sshd-defaults` does exist,
2. And if `dpkg-divert --truename /etc/pam.d/sshd` is not `/etc/pam.d/sshd-defaults`
3. We run `dpkg-divert --no-rename --divert /etc/pam.d/sshd-defaults --add /etc/pam.d/sshd`

Either path, the point is that `dpkg-divert --truename /etc/pam.d/sshd` is `/etc/pam.d/sshd-defaults` and `sshd-defaults` exists, everywhere and in an hour, and then we're ready for phase two.

I still don't get what it means to have a truename diverted, but this should do it everywhere, either (I assume?) creating sshd-defaults with `--rename` or running `--no-rename` if ssh-defaults is already there.

daaang replied:
Is this to catch dpkg up to what we're currently doing with puppet, so that we can have puppet let go of the files and undo this with dpkg so that it takes over when we finish?

rrotter replied:
- `dpkg-divert --divert` tells dpkg that when it's managing a file that a package expects to be at `/etc/pam.d/sshd`, that it should be at `/etc/pam.d/sshd-defaults` instead.
- `--rename` is the option you'd normally want to use. This says: rename sshd to sshd-defaults. This is great as long as `/etc/pam.d/sshd-defaults` does not exist, if it does, this command will always fail.
- With the `--no-rename` flag, we make the same changes to the dpkg database, but the file doesn't get moved. Then we have to follow up and ensure the correct data is at `/etc/pam.d/sshd-defaults`.
- `dpkg-divert --truename /etc/pam.d/sshd` prints the path where dpkg expects to find `/etc/pam.d/sshd`. If it prints `/etc/pam.d/sshd-defaults` we've already successfully diverted the file, and we don't need to do it again.

rrotter replied:
The goal is that dpkg should manage `sshd-defaults` in leiu of managing `sshd`, which we'll continue managing with puppet.

Once we remediate existing hosts we'll only need `dpkg-divert` to run on new hosts, where it will create `sshd-defaults` by relocating `sshd`, so puppet won't need to write this file anymore.

We have three starting states we need to handle correctly:
1. new host
2. existing host with puppet managed `sshd` and `sshd-defaults`
3. existing host where this code has already run

In each case the end state should be:
* `pam.d/sshd-defaults` contains the data that Debian normally writes to `pam.d/sshd`
* dpkg has "diverted" `pam.d/sshd` accordingly
* `pam.d/sshd` has the same contents we're already generating, but order is critical, because we can't have this data end up in sshd-defaults instead

daaang reviewed (commented):
Overall, I see what this is doing. But I do not understand what the divert does, and I do not understand what you want it to do, so I cannot say whether this will do what you want it to do.

rrotter committed:
update global ssh_config settings

This template previously replaced the debian ssh_config, now we're
writing it to a drop-in where it only overrides debian settings.

- don't set SendEnv, debian defaults are fine
- do set HashKnownHosts, debian defaults are not fine

rrotter commented:
rebased to production.
